addons/account/models/account_move.py

```diff
@@ -2549,7 +2549,12 @@ def _autopost_draft_entries(self):
             ('date', '<=', fields.Date.today()),
             ('auto_post', '=', True),
         ])
-        records.post()
+        if records:
+            cr = self.env.cr
+            for rec in records :
+                rec.post()
```

Comment:
You are probably disabling optimization related to pre-fetching; please use batches of 1000 instead.
And remove the extra space before the `:`.

Reply:
You can use :
```
for recs in self.env.cr.split_for_in_conditions(records):
      recs.post()
      cr.commit()
```
it split every 1000 (by default)

Reply:
@william-andre Hi
need your help
batches can we use in two styles.
1. while loop and give limit in the search method (multiple time connection with DB)
2. search all the record and after split the records (one-time connection with DB)
So which is better here?

Reply:
@fmdl Thank you for the info.

Reply:
@william-andre Hi again,
We also test that how many moves are posted in one cron job.
So the result is 1600 in one cron job.
So I think we need to use batches of 500 or 100, what you think?

because in 1000 batches almost 60% process is completed in the second batch.

Reply:
Why not 
```
batch_size = ...
records = self.search([
    ('state', '=', 'draft'),
    ('date', '<=', fields.Date.today()),
    ('auto_post', '=', True),
])
records[:batch_size].post()
```

Reply:
I just tried to run this test
```python
    def test_perfs(self):
        import time
        start = time.time()
        account_id = self.env['account.account'].search([('company_id', '=', self.env.company.id)], limit=1).id
        moves = self.env['account.move'].create([{
            'type': 'entry',
            'line_ids': [(0, 0, {'account_id': account_id, 'name': 'line'})],
        } for i in range(1000)])
        created = time.time()
        moves.action_post()
        posted = time.time()
        print(created-start, posted-created, posted-start)
```

Here is what it prints
```
3.6165683269500732 8.414510726928711 12.031079053878784
```

How do you have a timeout after 1600 records? What is taking up so much time? Is it an EDI if some kind or another external service?
If I'm not mistaken, the timeout of a cron on the SaaS is 45 minutes...

Reply:
cron job timeout after 15 minutes in sh

Reply:
> cron job timeout after 15 minutes in sh
That still means I can post `15×60÷8,42×1000 = 106888` in this period before having the timeout...

Reply:
let me check same in DB

Reply:
This is my test code run in DB
```
def test_perfs(self):
    import time
    start = time.time()
    records = self.env['account.move'].search([
            ('state', '=', 'draft'),
            ('date', '<=', fields.Date.today()),
            ('auto_post', '=', True),
        ], limit=1000)
    line_len = len(records.line_ids)
    records.post()
    posted = time.time()
    print(line_len, posted-start)
```

and print result is `2000 127.41367268562317`
so 15×60÷127.41×1000 = 7063.80

Reply:
We are run normally in migration server to test

`Batch >  1 2000 173.2004678249359`
`2021-02-22 10:36:06,096 598096 WARNING 7329 odoo.service.server: Thread <Thread(odoo.service.http.request.140305591154432, started 140305591154432)> virtual real time limit (177/120s) reached.`

Reply:
> Why not
> 
> ```
> batch_size = ...
> records = self.search([
>     ('state', '=', 'draft'),
>     ('date', '<=', fields.Date.today()),
>     ('auto_post', '=', True),
> ])
> records[:batch_size].post()
> ```

@smetl  Using this it's processed in batch or only take a first batch_size(1000) records?